[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

changpeng opened the pull request:
If the parameters(the input and segment select) coming in to amdgcn.trig.preop intrinsic are compile time constants, we pre-compute the output of amdgcn.trig.preop on the CPU and replaces the uses with the computed constant.

  This work extends the patch https://reviews.llvm.org/D120150 to make it a complete coverage.

  For the segment select, only src1[4:0] are used. A segment select is invalid if we are selecting the 53-bit segment beyond the [1200:0] range of the 2/PI table. 0 is returned when a segment select is not valid.

changpeng committed:
[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

  If the parameters(the input and segment select) coming in to
amdgcn.trig.preop intrinsic are compile time constants, we
pre-compute the output of amdgcn.trig.preop on the CPU and
replaces the uses with the computed constant.

  This work extends the patch https://reviews.llvm.org/D120150 to
make it a complete coverage.

  For the segment select, only src1[4:0] are used. A segment slect is
invalid if we are selecting the 53-bit segment beyond the [1200:0] range
of the 2/PI table. 0 is returned when a segment select is not valid.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+        0x7527bac7, 0xebe5f17b, 0x3d0739f7, 0x8a5292ea, 0x6bfb5fb1, 0x1f8d5d08,
+        0x56033046};
+
+    const APInt &SegVal = Cseg->getValue();
```
Can you apply the value clamp here instead of nested in the expression below

changpeng replied:
> Can you apply the value clamp here instead of nested in the expression below

Use "unsigned" for Shift, so this is no longer applicable.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+
+    const APInt &SegVal = Cseg->getValue();
+    bool Ovflow = false;
+    unsigned Numbits = 32;
```
Capitalize bits

changpeng replied:
> Capitalize bits

Use "32" directly in the current revision

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    APInt Bshift = Shift & 0x1f;
+    Numbits = 64;
+    Signed = false;
+    uint64_t Hi = ((uint64_t)TwoByPi[Idx] << 32) | (uint64_t)TwoByPi[Idx + 1];
```
Can use Make_64 for this

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    }
+
+    APInt Bshift = Shift & 0x1f;
+    Numbits = 64;
```
Mutating this is confusing. This no longer has the same use as above. Why is this part performed in 64-bit?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    bool Ovflow = false;
+    unsigned Numbits = 32;
+    bool Signed = true;
+    APInt EClamp(Numbits, 1077, Signed);
```
APSInt?

changpeng replied:
> APSInt?

Use "1077" directly. No need to create an APInt or APSInt

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    E &= 0x7ff;
+    E = E.trunc(Numbits);
+    APInt Shift =
+        (E.sgt(EClamp) ? E.ssub_ov(EClamp, Ovflow) : APInt(Numbits, 0, Signed))
```
APSInt

jayfoad replied:
All of the calculation of Shift and Idx should be done with int or unsigned. There's no need to use APInt or APSInt. E.g. you can start with `unsigned SegVal = Cseg->getValue().trunc(5).getZExtValue()`.

changpeng replied:
> All of the calculation of Shift and Idx should be done with int or unsigned. There's no need to use APInt or APSInt. E.g. you can start with `unsigned SegVal = Cseg->getValue().trunc(5).getZExtValue()`.

Done! Nice suggestions, Thanks.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,97 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+                     Ovflow);
+    uint32_t Idx = Shift.ashr(5).getZExtValue();
+
+    // Return 0 for invalid segment select (outbound).
```
Does this match the hardware behavior? The pseudocode in the manual does a bounds check for exponent > 1077 shift += exponent - 1077

changpeng replied:
We follow the hardware behavior exactly. The second revision is more clear.

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:
```diff
@@ -5723,84 +5743,80 @@ define double @trig_preop_constfold_strictfp() strictfp {
+  %val = call double @llvm.amdgcn.trig.preop.f64(double 0x7B00000000000000, i32 5)
   ret double %val
 }
 
```
Needs more test values

changpeng replied:
> Needs more test values

Would you please tell what are these more test values you expect? The existing tests cover
1. exponent of src0:  <= 1077, in between 1077 and 1968, and >=1968
2. segment (src1): negative, inbound and outbound

The only additional values I can think is first 52 bits are non zeros.

changpeng replied:
any additional comments?

arsenm replied:
those exponents with different mantissas, particularly the all 0/all 1 cases. And some denormals

changpeng replied:
Done! Now tests cover all 0s, all 1s and an arbitrary 52bit for mantissas for each exponent - segment combination.

changpeng committed:
[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

  If the parameters(the input and segment select) coming in to
amdgcn.trig.preop intrinsic are compile time constants, we pre-compute
the output of amdgcn.trig.preop on the CPU and replaces the uses with
the computed constant.
  This work extends the patch https://reviews.llvm.org/D120150 to
make it a complete coverage.
  For the segment select, only src1[4:0] are used. A segment select is
invalid if we are selecting the 53-bit segment beyond the [1200:0] range
of the 2/PI table. 0 is returned when a segment select is not valid.

changpeng commented:
Any additional comments?  Thanks.

changpeng committed:
[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

  If the parameters(the input and segment select) coming in to
amdgcn.trig.preop intrinsic are compile time constants, we pre-compute
the output of amdgcn.trig.preop on the CPU and replaces the uses with
the computed constant.
  This work extends the patch https://reviews.llvm.org/D120150 to
make it a complete coverage.
  For the segment select, only src1[4:0] are used. A segment select is
invalid if we are selecting the 53-bit segment beyond the [1200:0] range
of the 2/PI table. 0 is returned when a segment select is not valid.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,84 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    if (!Cseg)
+      break;
+
+    APInt Exponent = Fsrc.bitcastToAPInt().ashr(52);
+    Exponent &= 0x7ff;
+    Exponent = Exponent.trunc(32);
```
```suggestion
    uint64_t Exponent = (Fsrc.bitcastToAPInt().getZExtValue() >> 52) & 0x7ff;
```

changpeng replied:
Done as suggested, thanks.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,84 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    }
+
+    unsigned BShift = Shift & 0x1f;
+    APInt Thi = APInt(64, Make_64(TwoByPi[Idx], TwoByPi[Idx + 1]), false);
+    APInt Tlo = APInt(64, Make_64(TwoByPi[Idx + 2], 0), false);
```
Use uint64_t instead of APInt?

changpeng replied:
Okay to use uint64_t. Done!

changpeng committed:
[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,81 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    }
+
+    unsigned BShift = Shift & 0x1f;
+    uint64_t Thi = ((uint64_t)TwoByPi[Idx] << 32) | (uint64_t)TwoByPi[Idx + 1];
+    uint64_t Tlo = (uint64_t)TwoByPi[Idx + 2] << 32;
```
You can still use `Make_64` here.

changpeng replied:
right! Thanks

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,81 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+    if (!Cseg)
+      break;
+
+    uint64_t Exponent = (Fsrc.bitcastToAPInt().getZExtValue() >> 52) & 0x7ff;
```
Nit: might look neater as `unsigned` since you know the value is <= 0x7ff so it will fit.

changpeng replied:
Right!

jayfoad commented on llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:
```diff
@@ -5707,10 +5701,36 @@ define double @trig_preop_variable_args(double %arg0, i32 %arg1) {
 ; CHECK-LABEL: @trig_preop_constfold(
-; CHECK-NEXT:    [[VAL:%.*]] = call double @llvm.amdgcn.trig.preop.f64(double 3.454350e+02, i32 5)
-; CHECK-NEXT:    ret double [[VAL]]
+; CHECK-NEXT:    ret double 0x394A6EE06DB14ACC
```
Do you have some way of checking that this matches the hardware result, for a few randomly chosen inputs?

changpeng replied:
Yes, verified with an OpenCL program using the builtin to make the outputs.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,81 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+
+    const APFloat &Fsrc = Csrc->getValueAPF();
+    if (Fsrc.isNaN()) {
+      // FIXME: We just need to make the nan quiet here, but that's unavailable
+      // on APFloat, only IEEEfloat
+      auto *Quieted = ConstantFP::get(
+          II.getType(), scalbn(Fsrc, 0, APFloat::rmNearestTiesToEven));
```
This fixme was fixed, you should be able to do just Fsrc.makeQuiet

changpeng replied:
Done, Thanks.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1102,6 +1102,81 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+
+    Value *Src = II.getArgOperand(0);
+    Value *Segment = II.getArgOperand(1);
+    if (isa<PoisonValue>(Src))
```
If segment is poison, the result should also be poison?

changpeng committed:
[AMDGPU] Constant folding of llvm.amdgcn.trig.preop

arsenm reviewed (approved):
LGTM. I would feel better if we had some conformance testing with constant folding values, but we don't

rampitec commented:
As a side note I'd like to see who and why calls any trig function on a constant argument exceeding 360 grads.
